storage/store.go

```diff
@@ -175,42 +193,27 @@ func (e *NotBootstrappedError) Error() string {
 	return "store has not been bootstrapped"
 }
 
-// storeRangeIterator is an implementation of rangeIterator which
-// cycles through a store's rangesByKey slice.
-type storeRangeIterator struct {
-	store     *Store
-	remaining int
-	index     int
-}
-
-func newStoreRangeIterator(store *Store) *storeRangeIterator {
-	r := &storeRangeIterator{
-		store: store,
-	}
-	r.Reset()
-	return r
+// storeRangeSet is an implementation of rangeSet which
+// cycles through a store's rangesByKey btree with mutex locking.
+type storeRangeSet struct {
+	store *Store
 }
 
-func (si *storeRangeIterator) Next() *Range {
-	si.store.mu.Lock()
-	defer si.store.mu.Unlock()
-	if index, remaining := si.index, len(si.store.rangesByKey)-si.index; remaining > 0 {
-		si.index++
-		si.remaining = remaining - 1
-		return si.store.rangesByKey[index]
-	}
-	return nil
-}
-
-func (si *storeRangeIterator) EstimatedCount() int {
-	return si.remaining
+func (rs *storeRangeSet) Visit(iterator func(*Range) bool) {
+	// Hold the lock except when the iterator is called.
```

Comment:
In order to be satisfying, such a test would need to be a fairly expensive brute force process (we don't want to make too many assumptions about things like when internal nodes are split, since the point of the test is to catch any changes that might be introduced upstream). I'd rather see a statement from upstream about what guarantees they intend to provide before relying on anything here. (for example, this [c++ btree](https://code.google.com/p/cpp-btree/wiki/UsageInstructions) has both safe and unsafe variants)

And even if we verified the current behavior, there's a question of whether that current behavior is acceptable (any change may cause large chunks of the tree to be skipped, so we need an entire scanner interval to pass without changes to be sure we've seen everything). I'd start by making a copy, and if we can't afford that we could consider adding a "safe" iterator variant like that c++ implementation that has better behavior when things change out from under it.